src/main/java/com/couchbase/lite/Database.java

```diff
@@ -1152,27 +1149,54 @@ private void markBlobstoreMigrated() {
      */
     @InterfaceAudience.Private
     public boolean close() {
-        if(!open) {
+        if (!open) {
             return false;
         }
 
-        if(views != null) {
+        if (views != null) {
             for (View view : views.values()) {
                 view.databaseClosing();
             }
         }
         views = null;
 
-        if(activeReplicators != null) {
-            for(Replication replicator : activeReplicators) {
+        if (activeReplicators != null) {
+            List<CountDownLatch> latches = new ArrayList<CountDownLatch>();
+            for (Replication replicator : activeReplicators) {
+                // handler to check if the replicator stopped
+                final CountDownLatch latch = new CountDownLatch(1);
+                replicator.addChangeListener(new Replication.ChangeListener() {
+                    @Override
+                    public void changed(Replication.ChangeEvent event) {
+                        if (event.getSource().getStatus() == Replication.ReplicationStatus.REPLICATION_STOPPED) {
+                            latch.countDown();
+                        }
+                    }
+                });
+                latches.add(latch);
+
+                // ask replicator to stop
                 replicator.databaseClosing();
             }
+
```

Comment:
1. As I don't fully understand the threading model of the java replicators, I assume that this code snippet is thread-safe (e.g. it's not possible that a replicator stopped while adding a change listener to that replicator). Can you confirm that?

2. As we are going to wait until all of the replicators to be done, it's really unknown to me that this is going to pause the app if I call database.close() in my app. What is your opinion on this? Honestly I need to try this with the CBL iOS as well to see the current behavior.

I don't fully understand the replicator stopping chains. @tleyden can you review this part as well?